pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -517,12 +568,66 @@ func FindRegistry(ctx *types.SystemContext, ref string) (*Registry, error) {
 	return nil, nil
 }
 
-// tomlConfig loads the unmarshals the configuration at the specified path.
-// Note that the tomlConfig's fields will be overridden if they are set in
-// specified path.  This behavior is necessary for registries.conf.d stances
-// behavior.
-func (t *tomlConfig) loadConfig(path string) error {
+// tomlConfig loads and unmarshals the configuration at the specified path. Note
+// that v1 configs are translated into v2 and are cleared.  Use forceV2 if the
+// config must in the v2 format.
+//
+// Note that specified fields in path will replace already set fields in the
+// tomlConfig.  Only the [[registry]] tables are merged by prefix.
+func (c *tomlConfig) loadConfig(path string, forceV2 bool) error {
 	logrus.Debugf("Loading registries.conf %q", path)
-	_, err := toml.DecodeFile(path, t)
-	return err
+
+	// Save the registries before decoding the file where they could be lost.
+	// We merge them later again.
+	registryMap := make(map[string]Registry)
+	for i := range c.Registries {
+		registryMap[c.Registries[i].Prefix] = c.Registries[i]
+	}
+
+	// Load the tomlConfig. Note that `DecodeFile` will overwrite set fields.
+	c.Registries = nil // important to clear the memory to prevent us from overlapping fields
```

Comment:
FIXME:

Save-clear-unmarshal-process-overwrite feels a bit heavy-handed. Why not, for each file, unmarshall-process in a new object, and then merge? That would give us complete control over the merging policy.

OTOH risk of forgetting to merge new fields?

Reply:
> Why not, for each file, unmarshall-process in a new object, and then merge?

Because otherwise we cannot distinguish between `unqualified-search-registies = []` and them  not being set at all. `c.Registries = nil` is needed as the "old" registries and the "new" registries start at the same memory address which could lead to unintended overlaps.

Reply:
> Because otherwise we cannot distinguish between `unqualified-search-registies = []` and them not being set at all.

AFAICS unset is `nil`, set to empty is `[]string{}`.

OTOH this is less code, it does work fine, and anyway it’s private.